### ISSUE TITLE
backward(create_graph=True) should raise a warning for potential memory leak

### ISSUE THREAD

**Issue by benvcutilli** [pytorch/pytorch#4661]
Ok, sorry to bother you again, and really sorry if I am just doing it wrong again (I looked at the docs for things that would address this, but didn't see anything after looking at it quickly).

It appears that the next segment of code leaks CUDA memory. I don't believe I had this problem on 0.2.0 (or some version close to that) on the code that I was running (that I prefer to not publicly disclose, which is why I came up with this example instead):

```
net = torch.nn.Linear(30000,1).cuda()
data = torch.ones(100, 30000).cuda()
for rep in range(1000):
  for batchIndices in range(0,100,100):
    batch = torch.autograd.Variable(data[batchIndices:batchIndices+10], requires_grad=True)
    net.forward(batch).norm(2).backward(create_graph=True)
    batch.grad.norm(2).backward()
```

`nvidia-smi` output:

After first run of this:
```
|   0  TITAN X (Pascal)    Off  | 00000000:02:00.0 Off |                  N/A |
| 33%   55C    P8    14W / 250W |   2417MiB / 12189MiB |      0%      Default |
```

After second run of this (but without the lines defining `net` and `data`):
```
|   0  TITAN X (Pascal)    Off  | 00000000:02:00.0 Off |                  N/A |
| 32%   51C    P2    58W / 250W |   4417MiB / 12189MiB |      0%      Default |
```

This seems like a similar issue to https://github.com/pytorch/pytorch/issues/3824. The code in that issue's first post that reproduces it is similar to mine, so I suspect I am doing it correctly (just to defend myself!).

`torch.__version__` is `'0.3.0.post4'`

I downgraded pytorch to 0.2.0 to make sure it didn't happen, and it appears not to; `torch.__version__` says `'0.2.0_0'`.

Sorry if you are working on this already; I hope that some more feedback wouldn't be too annoying, at least.

cc @ezyang @SsnL @albanD @zou3519 @gqchen @ngimel

**Comment by benvcutilli** [comment]
I noticed I have some indices wrong, but I don't think it would invalidate the bug (the "skip 100" in one of the ranges vs the "+10" in the slicing)

**Comment by apaszke** [comment]
Can you please update the snippet (you can edit your posts)?

**Comment by benvcutilli** [comment]
The code I ran was the one I originally posted (I just copy-pasted it from my terminal). My other comment was just acknowledging that I made an indexing mistake, but that it most likely wouldn't change the fact that it is (allegedly) leaking; it would just repeatedly access the same section of 10 of `data` instead of going by 10s through a 100-sample dataset.

**Comment by soumith** [comment]
thanks a lot for the snippet Ben, we'll look into it. if it's not leaking with v0.2.0 and leaking with 0.3.0, it's worth investigating.

**Comment by soumith** [comment]
i'm able to reproduce the issue on `master`. Indeed it looks like a memory leak.
I've tried `import gc; gc.collect()` and also `torch.cuda.empty_cache()` as well.

**Comment by soumith** [comment]
A more simplified test case:

```
import torch
net = torch.nn.Linear(30000,1).cuda()
data = torch.ones(100, 30000).cuda()
for rep in range(1000):
  for batchIndices in range(0,100,100):
    batch = torch.autograd.Variable(data[batchIndices:batchIndices+10], requires_grad=True)
    net.forward(batch).norm(2).backward(create_graph=True)
```

**Comment by benvcutilli** [comment]
Cool, happy that it was reproducible. I'm away from my machine right now,
so I can't check this myself, but I think I might have been able to get it
to stop leaking by having `data` cpu-only, `Variable`-ing the batch, then
converting that Variable to `cuda`. Currently, the conversion to `Variable`
happens after `data` is in cuda form. I'm not sure if that worked or not,
to be clear, but I think it might've, if that helps.

I will check this when I get back, but it won't be for a few hours, so
anyone obviously take a look first if they are itching to.

Le 14 janv. 2018 09:45, "Soumith Chintala" <notifications@github.com> a
écrit :

**Comment by benvcutilli** [comment]
Ok, just checked it, and I was wrong. Still leaks if you modify it the way I said to in the last comment (I used Soumith's "simplified" snippet).

**Comment by albanD** [comment]
@soumith  cuda is not necessary to get a leak as your sample without it will leak cpu memory.
Also after investigation, no python object is leaked (checked using debug build). So this is most likely on the cpp side.

**Comment by benvcutilli** [comment]
@soumith @apaszke if it helps, this even simpler code leaks as well:

```
x = torch.autograd.Variable(torch.ones(1).cuda(), requires_grad=True)
for rep in range(1000000):
  (x*x).backward(create_graph=True)
```

**Comment by apaszke** [comment]
Oh yeah, that's actually a known thing. We discussed this, but I can't remember what the final decision was. @colesbury?

**Comment by colesbury** [comment]
I think we said we were going to make `.grad` never store history. So if you want to to higher order differentiation, you'd want to use `torch.autograd.grad` instead of `.backward()`

**Comment by benvcutilli** [comment]
Why wasn't this an issue in previous versions? Was some change made so that doing what was done before in those versions is not possible anymore?

**Comment by benvcutilli** [comment]
I put the creation of the variable in the loop (to make it consistent with my first snippet), and still leaks (even faster):

```
for rep in range(1000000):
  x = torch.autograd.Variable(torch.ones(1).cuda(), requires_grad=True)
  (x*x).backward(create_graph=True)
```

Also, I realize that there may be some tough design decisions when it comes to this kind of stuff, but changing the paradigm from Variable-oriented calling to `torch.autograd`-module-oriented calling would be very confusing and not intuitive. Just something to consider.

**Comment by apaszke** [comment]
Yeah we changed one weak reference in the C++ code to a strong reference and that's what's causing the problems.

**Comment by benvcutilli** [comment]
Maybe I'm not thinking of the computational graph correctly, but in my second snippet, shouldn't any strong reference related to `x` be broken when `x` is re-assigned, thereby allowing garbage collection?

**Comment by apaszke** [comment]
No because once you create a cycle of strong references your only chance is getting Python GC to detect it's unreachable. However, this cycles is entirely in our C++ backend, so it's invisible to the GC.

**Comment by benvcutilli** [comment]
@colesbury @apaszke FWIW using torch.autograd.grad() instead isn't leaking in my **super top secret** code, as you predicted. Is there any way `backward()` for `.grad`s could be used as a wrapper for `torch.autograd.grad()` to keep the object semantics the same no matter what `n`th derivative you're calculating? For instance, there could be a rare case, say, if you were taking the Taylor expansion where higher- and higher-order derivatives of the gradient or whatever (I know, this is very contrived; sort of inspired by Newton-Raphson but in a very very abstract way), and in order to do that you would write a loop that calls `.backward()`. Otherwise, you'd have a conditional for which order derivative you're calculating and you make the corresponding call.

I mean, you could always just call with `torch.autograd.grad()` for all of them, but, I dunno, hopefully this illustrates the general point of consistency. The `.backward()` semantic is really nice, too. Don't know if this is a computationally tractable request at this point, though.

**Comment by apaszke** [comment]
I'm not sure what you mean, can you pelase elaborate?. `.backward()` isn't very well suited for higher order derivatives, mostly because if you do `.backward()` twice, you'll get a sum of both results (so e.g. a sum of first and second order derivative).

**Comment by benvcutilli** [comment]
Yeah, I understand that, and agree that that could be confusing (I just zero out the gradients, but I suppose that someone who doesn't have a good understanding of how PyTorch works could be confused by that; I'd have to think about that a little to make that more intuitive). I like the `.backward()` semantics because it is easy to read and easy to program. For example, as long as it isn't too verbose, you could rifle off something like this, which takes one line:

```
(batch.grad).pow(2).sum().backward()
```

where there is a clear flow from left-to-right. This is nice for the programmer and the reader. Of course, you can also just do, as you and @colesbury suggested (not sure if my parameters are ordered correctly -- which may be another downside of the following method, unless `kwargs` are accepted),

```
torch.autograd.grad( variableToOptimize, (batch.grad).pow(2).sum() )
```

and the programmer understands what is going on, but the person who wrote it will take a little longer to understand because you've switched from OOP to regular function calling within the same line. In a math-intensive environment, I feel like the confusion of this is amplified as it is already difficult enough to keep track of every variable and function that is flying around in the code.

If you could do something like `.backward(input_variables=(something, somethingElse))`, that would be in the spirit of `torch.autograd.grad()` but keep the OOP properties of `.backward()`.

**Comment by benvcutilli** [comment]
Maybe `.backward()` would change, to again "be in the spirit of `torch.autograd.grad()`", to return the gradient instead of "accumulate" them in the leaves. Or have some option specifying which mode you want.

**Comment by youkaichao** [comment]
The same problem makes me confused for days. I want to work with high order derivatives and `.backward(create_graph=True)` incurs memory leak. Switching to `autograd.grad` solves the problem.

**Comment by IsThatYou** [comment]
Recently I need to double backprop on the gradients of the embedding layers for NLP tasks. This requires me to retrieve gradients of the embedding layer before they get added up (if there are multiple tokens in the same sentence, then `autograd.grad` return the same of the gradients). Hence I have to use backward_hooks to get the gradient. This requires me to either call `.backward(create_graph=True)` or `autograd.grad` so that the hook can capture the gradients. Both have a memory leak issue if I call backward on the gradients captured by the hooks.

I am not sure if this is caused by the same issue? Should I open a new issue for this?

**Comment by albanD** [comment]
Hi,

In general, if you have questions, please use the forum: https://discuss.pytorch.org/
We only use issues for feature requests or bugs.

**Comment by IsThatYou** [comment]
Hi,
Thank you for the response. I am fairly certain this is caused by a bug. If I just use `autograd.grad` to get the gradient with respect to the embedding layer, let's call this W', then I can call backward on W' without any memory leak. The problem is that this is not exactly what I wanted. It could be off because if there are the same tokens in the sentence, the gradient returned for that token would be the sum of the gradients of that token in multiple occurrences. So the only way to do it, in this case, would be to backward on the gradients captured by a hook. This is achieved by setting a hook at the embedding layer, and then call either `.backward(create_graph=True)` or `autograd.grad(embedding_layer, loss, create_graph=True)`. Then the gradients would be captured by the hook and they are actually separated rather than summed up for the same tokens. However, this incurs a memory leak of the GPU. 

I will also post on the forum at the same time. But I can produce this pretty consistently on a BERT and a CNN model. 

Here is like a minimized code that causes it:
```
import torch
import os
import numpy as np
np.random.seed(0)
torch.manual_seed(0)
os.environ['CUDA_VISIBLE_DEVICES']="1"
dev = torch.device('cuda')

embedding_gradient = []
def hook_layer(module, grad_in, grad_out):
    embedding_gradient.append(grad_out[0])
arr = np.random.randint(10, size=3)
random_ix = torch.LongTensor(arr).to(dev)
print(arr)
embedding_layer = torch.nn.Embedding(10, 5, padding_idx=0).to(dev)
net = torch.nn.Linear(5,1).to(dev)
for i in range(1000):
    print(i)
    print(torch.cuda.memory_summary(device=0, abbreviated=True))
    #1 set the hook to embedding layer
    embedding_gradient = []
    hook = embedding_layer.register_backward_hook(hook_layer)
    #2 forward pass
    embeds = embedding_layer(random_ix)
    out = net(embeds)
    #3 backward pass
    summed = out.norm(2)
    summed.backward(create_graph=True)
    #4 remove the hook
    hook.remove()
    final = embedding_gradient[0].sum()
    final.backward()
```
It is interesting that in the above code snipet, if I use `grad_auto = torch.autograd.grad(summed, embedding_layer.weight, create_graph=True)` rather than `summed.backward(create_graph=True)`, the memory leak goes away. But it is not the case in my actual code. I am still struggling to reproduce it.

**Comment by albanD** [comment]
We should be clearer in the doc that `.bacward(..., create_graph=True)` creates reference cycle and should either be replaced by `autograd.grad()` or make sure the user manually deletes the `.grad` after use.

**Comment by Juna2** [comment]
> We should be clearer in the doc that `.bacward(..., create_graph=True)` creates reference cycle and should either be replaced by `autograd.grad()` or make sure the user manually deletes the `.grad` after use.

Sorry but I have one more question.
You said "manually deletes the .grad after use"
Is it okay to do like this then?
```
loss.backward(create_graph=True)
model.zero_grad()
```

**Comment by albanD** [comment]
No this only sets the content to 0. You want to remove them by doing setting them to None or calling del:
```python
for p in model.parameters():
    p.grad = None
```

**Comment by benvcutilli** [comment]
With a due respect (and appreciation for what you all have done by building this framework), I really don’t think hoping that the user will see the note in the docs or requiring them to basically do memory management in Python makes any sense. This should just be fixed.

**Comment by albanD** [comment]
Hi,

I think the main reason for backward to exist is to seamlessly operate with the other torch constructs. In particular with `nn.Parameter`, `nn.Module` and `torch.optim`.
The goal being that the user can write a training loop without having to manipulate all the gradients by hand.

And because of this, I don't think we want to change the current behavior of `.backward()` to populate the `.grad` field of Tensors it computes gradients for.
Unfortunately, the issue that you raised here is directly linked to the existence of this `.grad` field.
This is why we did not changed the behavior and just warn the user that what they do can be dangerous if they don't use it properly.
Does that make sense?

**Comment by benvcutilli** [comment]
I get that reasoning, but I don’t think it makes sense to leave it as-is because it is clearly a bug, and I’ve rarely ever heard good reasons as to why things are left as-is.

Maybe you should change the behavior of .backward(...) so that it behaves in a traditional manner when calculating the first derivative, and then it behaves the way torch.autograd.grad(...) does otherwise.

**Comment by albanD** [comment]
> it is clearly a bug

I would agree with you. But this is one we cannot solve. The reference cycle here is too complex between both python and c++ objects to be able to be broken (in any fast manner).

From my experience, there are valid use cases where people use `.backward(create_graph=True)` so we cannot prevent users from doing this.
Also having `.backward()` not accumulate into the `.grad` field would be very surprising I think as it is the one thing this API exists for. 

We could be more aggressive here and raise warning when we detect bad behavior from the users. But I can't think of a way to do so: `.backward(create_graph=True)`  in itself is not a problem. Accumulate gradients into a `.grad` that requires_grad isn't either (your example above does that and is doing the accumulation as expected). Do you have other ideas where we could raise warning that would not spam users doing valid things?

**Comment by benvcutilli** [comment]
Well, the reason why I suggested that you change the behavior to a wrapper when computing successive derivatives is that, as far as I know, anyone who is using create_graph plans to do second- or higher-order differentiation. So if you don’t fix it, they’re likely going to run into memory leaks, which is a show-stopper unless they only run things for short amounts of time with little data. If you do fix it via the manner I propose, yeah their code is gonna break, but it was going to be broken anyway.

**Comment by SsnL** [comment]
My 2 cents. I've done higher-order differentiation many times. I've always used `autograd.grad` instead of `.backward` or `autograd.backward` because 
1. `autograd.grad` doesn't accumulate gradient with previous values, which can lead to obscure errors.
2. it is almost always clearer in code than accessing `.grad` of leaves.
3. it avoids ref cycle.

In fact, most DL uses of higher-order gradients I've seen use `autograd.grad`. So could we add a warning (either in doc or in code) to `backward(create_graph)` and advocate `autograd.grad` instead?

**Comment by albanD** [comment]
The warning is already in the doc (in master, not in 1.5).
Adding a warning to every call to `backward(create_graph)` might be a bit verbose but I agree with the argument that `autograd.grad` should be used anyway in most cases (I can think of cases where you use (Distributed)DataParallel or checkpointing where using .backward() is mandatory). But I guess a warn_once will be fine and people that really want this can just silence the warning.

**Comment by benvcutilli** [comment]
Just in case you would like to point to a reference, this tweet by Zachary Lipton on 2018.8.30, https://twitter.com/zacharylipton/status/1167412117647089664?s=21 (via some Twitter recommendation feature) is why I brought up the argument that documentation is not read by everyone (even though it should be), so making a comment about this bug in the documentation alone isn’t a good idea. The warning is better, but IMO not the best idea (I’ve already stated what I think should be done). The warning should at least sleep the program for enough seconds so that the warning can be noticed and read; the reason why I bring this up is because I often print out loss values during learning, so the warning would likely fly off the screen as soon as it’s displayed.

**Comment by albanD** [comment]
Note that because the `Tensor.backward()` is directly hooked into the c++ API, this warning should be added directly [there](https://github.com/pytorch/pytorch/blob/5640b79bf8a5412a0209a919c05c811d5427cc12/torch/csrc/autograd/autograd.cpp#L131) and use the `TORCH_WARN_ONCE()` macro to generate the warning. It should also be tested in tedt/test_autograd.py.
